File: Var10JS/WaterBottles.js
const readline = require('readline'); // Импорт модуля readline для взаимодействия с консолью

const rl = readline.createInterface({ // Создание интерфейса для ввода-вывода
    input: process.stdin, // Ввод из стандартного потока ввода (консоль)
    output: process.stdout // Вывод в стандартный поток вывода (консоль)
});

function main() {
    rl.question("Введите количество банок с водой: ", (bottlesInput) => { // Запрос ввода количества банок с водой
        const bottles = parseInt(bottlesInput); // Преобразование введенной строки в число

        rl.question("Введите курс обмена пустых банок на новые: ", (rateInput) => { // Запрос ввода курса обмена
            const rate = parseInt(rateInput); // Преобразование введенной строки в число

            const bottlesDrank = drinkWater(bottles, rate); // Вызов функции для вычисления максимального количества выпитых банок
            console.log(`Максимальное число банок, которое можно выпить: ${bottlesDrank}`); // Вывод результата в консоль
            rl.close(); // Закрытие интерфейса ввода-вывода
        });
    });
}

function drinkWater(bottles, rate) {
    let bottlesDrank = bottles; // Инициализация переменной для хранения количества выпитых банок
    let emptyBottles = bottles; // Инициализация переменной для хранения количества пустых банок
    let steps = 1; // Инициализация переменной для подсчета шагов

    console.log("Дано: "); // Вывод сообщения о начале работы программы
    printHappyFaces(bottles); // Вывод изображения смайликов в консоль
    console.log("Шаг: " + steps); // Вывод текущего шага в консоль
    printEmptyBottles(emptyBottles); // Вывод изображения пустых банок в консоль

    while (emptyBottles >= rate) { // Пока количество пустых банок больше или равно курсу обмена
        const exchangedBottles = Math.floor(emptyBottles / rate); // Вычисление количества обмененных банок
        bottlesDrank += exchangedBottles; // Обновление общего количества выпитых банок
        const remainingEmptyBottles = emptyBottles % rate; // Вычисление оставшихся пустых банок
        emptyBottles = exchangedBottles + remainingEmptyBottles; // Обновление количества пустых банок
        steps++; // Увеличение счетчика шагов

        console.log("Шаг: " + steps); // Вывод текущего шага в консоль
        printHappyFaces(exchangedBottles); // Вывод изображения смайликов для обмененных банок в консоль
        printEmptyBottles(remainingEmptyBottles); // Вывод изображения пустых банок в консоль

        steps++; // Увеличение счетчика шагов
        console.log("Шаг: " + steps); // Вывод текущего шага в консоль
        printEmptyBottles(emptyBottles); // Вывод изображения пустых банок в консоль
    }

    console.log("Количество итераций: " + steps); // Вывод количества итераций в консоль
    return bottlesDrank; // Возврат общего количества выпитых банок
}

function printHappyFaces(n) {
    let happyFaces = '☺'.repeat(n); // Формирование строки смайликов
    if (n % 3 !== 0) happyFaces += '\n'; // Добавление переноса строки, если количество смайликов не кратно 3
    console.log(happyFaces); // Вывод строки смайликов в консоль
}

function printEmptyBottles(n) {
    let emptyBottles = '©'.repeat(n); // Формирование строки пустых банок
    if (n % 3 !== 0) emptyBottles += '\n'; // Добавление переноса строки, если количество пустых банок не кратно 3
    console.log(emptyBottles); // Вывод строки пустых банок в консоль
}

main(); // Вызов основной функции

File: Var10Python/WaterBottles.py
def main():
    bottles = int(input("Введите количество банок с водой: "))  # Ввод количества банок с водой
    rate = int(input("Введите курс обмена пустых банок на новые: "))  # Ввод курса обмена пустых банок на новые

    bottles_drank = drink_water(bottles, rate)  # Вызов функции для расчета максимального числа выпитых банок
    print(f"Максимальное число банок, которое можно выпить: {bottles_drank}")  # Вывод результата

def drink_water(bottles, rate):
    bottles_drank = bottles  # Инициализация переменной с количеством выпитых банок
    empty_bottles = bottles  # Инициализация переменной с количеством пустых банок
    steps = 1  # Инициализация переменной с количеством шагов

    print("Дано: ")  # Вывод заголовка
    print_happy_faces(bottles)  # Вызов функции для вывода смайликов
    print("Шаг:", steps)  # Вывод номера текущего шага
    print_empty_bottles(empty_bottles)  # Вызов функции для вывода пустых банок

    while empty_bottles >= rate:  # Проверка, есть ли еще пустые банки для обмена
        exchanged_bottles = empty_bottles // rate  # Вычисление количества банок для обмена
        bottles_drank += exchanged_bottles  # Увеличение количества выпитых банок
        remaining_empty_bottles = empty_bottles % rate  # Вычисление оставшихся пустых банок
        empty_bottles = exchanged_bottles + remaining_empty_bottles  # Обновление количества пустых банок
        steps += 1  # Увеличение числа шагов

        print("Шаг:", steps)  # Вывод номера текущего шага
        print_happy_faces(exchanged_bottles)  # Вызов функции для вывода смайликов
        print_empty_bottles(remaining_empty_bottles)  # Вызов функции для вывода пустых банок

        steps += 1  # Увеличение числа шагов
        print("Шаг:", steps)  # Вывод номера текущего шага
        print_empty_bottles(empty_bottles)  # Вызов функции для вывода пустых банок

    print("Количество итераций:", steps)  # Вывод количества итераций
    return bottles_drank  # Возвращаем количество выпитых банок

def print_happy_faces(n):
    happy_faces = "☺" * n  # Создание строки из смайликов
    if n % 3 != 0:  # Проверка, нужно ли перейти на новую строку
        happy_faces += '\n'  # Добавление перехода на новую строку
    print(happy_faces)  # Вывод смайликов

def print_empty_bottles(n):
    empty_bottles = "©" * n  # Создание строки из пустых банок
    if n % 3 != 0:  # Проверка, нужно ли перейти на новую строку
        empty_bottles += '\n'  # Добавление перехода на новую строку
    print(empty_bottles)  # Вывод пустых банок

main()  # Вызов основной функции программы
